File: lib/agent_blueprint_protocol/signature.ex
defmodule AgentBlueprintProtocol.Signature do
  @moduledoc """
  Detached JWS signature envelope: RFC 7515 compact serialization with the
  RFC 7797 `b64=false` unencoded, detached payload, Ed25519 signatures
  verified through `:crypto`. **Verify-only** — the package never signs,
  never accepts a private key on any function, and performs no key
  discovery or trust selection; hosts supply the trusted keys.

  What is signed is the RFC 7797 §3 input:

      ASCII(BASE64URL(JCS(protected_header))) || "." || JCS(signed_attributes)

  so a standard, `b64=false`-aware JOSE tooling stack can verify the same
  bytes (the ratified preimage form,
  byte-disjoint from the design's earlier sketch). **Producer contract:** the
  protected header segment must be the JCS serialization of the exact
  four-member header; a producer using any other member order produces a
  legitimately-signed JWS this verifier denies `:signature_not_verified`.
  This package is deliberately narrower than generic JOSE for determinism.

  The envelope shape (closed world, value-free denials):

      {"protected":        {"alg":"EdDSA","b64":false,"crit":["b64"],"kid":"…"},
       "signed_attributes":{"algorithm":"Ed25519","content_digest":"sha-256:…",
                            "created_at":"2026-08-20T00:00:00Z","key_id":"…",
                            "purpose":"blueprint"},
       "signature":        "<86-char unpadded base64url>"}

  `kid` must equal the signed `key_id` (key substitution prevention);
  `key_id` is dot-free so the unencoded payload can never contain `.`
  (RFC 7797 §5.2 forbids it in compact serializations); `created_at` is
  Z-form whole seconds. Attestations use the identical envelope with
  `purpose`/`content_digest` replaced by a registered `kind` and a
  `statement_digest`; the kind registry is empty until kinds become data,
  so every attestation denies `:attestation_malformed` today (fail-closed).

  Because `content_digest` and `purpose` are inside the signed bytes, a
  signature cannot be lifted onto a different artifact, digest, or
  purpose. Errors are value-free.
  Signature verification produces evidence only; it never authorizes an operation.
  """

  alias AgentBlueprintProtocol.{Base64Url, Canonicalization, Digest, Json}

  defmodule PublicKey do
    @moduledoc """
    A host-supplied Ed25519 public key, matched by the producer-chosen
    `key_id`. The package performs no discovery or trust selection.
    A public key is verification input that carries no authority.
    """

    @enforce_keys [:key_id, :algorithm, :key]
    defstruct [:key_id, :algorithm, :key]

    @type t :: %__MODULE__{key_id: binary(), algorithm: :ed25519, key: <<_::256>>}
  end

  defmodule Attributes do
    @moduledoc """
    The validated signed attributes of a signature entry, with
    `content_digest` parsed to a `Digest` value and `purpose` the closed
    purpose atom.
    Signed attributes are signature evidence that carries no authority.
    """

    @enforce_keys [:algorithm, :content_digest, :created_at, :key_id, :purpose]
    defstruct [:algorithm, :content_digest, :created_at, :key_id, :purpose]

    @type purpose :: :blueprint | :deployment | :federation_envelope

    @type t :: %__MODULE__{
            algorithm: binary(),
            content_digest: Digest.t(),
            created_at: binary(),
            key_id: binary(),
            purpose: purpose()
          }
  end

  @type reason ::
          :signature_algorithm_unsupported
          | :signature_key_unsupported
          | :signature_malformed
          | :signature_not_verified
          | :attestation_malformed

  @purposes %{
    "blueprint" => :blueprint,
    "deployment" => :deployment,
    "federation-envelope" => :federation_envelope
  }

  # The attestation kind registry: EMPTY by design — no kinds are
  # spec-named. Unknown kinds deny :attestation_malformed (fail-closed);
  # kinds become data when real consumers exist.
  @kinds %{}

  # RFC 3339 Z-form, whole seconds, no offset (deterministic JCS bytes; the
  # no-fraction rule also keeps the unencoded payload dot-and-dot-free).
  @created_at ~r/\A\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z\z/

  @entry_members ["protected", "signature", "signed_attributes"]
  @header_members ["alg", "b64", "crit", "kid"]
  @signature_attr_members ["algorithm", "content_digest", "created_at", "key_id", "purpose"]
  @attestation_attr_members ["algorithm", "created_at", "key_id", "kind", "statement_digest"]

  @doc """
  The RFC 7797 `b64=false` signing input of a signature entry (the tagged
  algebra as decoded by `Json`): `BASE64URL(JCS(protected)) || "." ||
  JCS(signed_attributes)`. Envelope violations deny with their
  `verify/2` reason.
  """
  @spec signing_input(Json.value()) :: {:ok, binary()} | {:error, reason()}
  def signing_input(entry) do
    with {:ok, parts} <- parts(entry, :signature) do
      {:ok, parts.header_b64 <> "." <> parts.payload}
    end
  end

  @doc """
  The signed attributes of a signature entry as an `Attributes` value.
  **Shape-validated only** — no signature has been checked; only `verify/2`
  establishes the signature fact. Do not consume attributes from an
  envelope that has not verified.
  """
  @spec attributes(Json.value()) :: {:ok, Attributes.t()} | {:error, reason()}
  def attributes(entry) do
    with {:ok, parts} <- parts(entry, :signature) do
      {:ok, parts.attributes}
    end
  end

  @doc """
  Verify a signature entry against host-supplied trusted keys. Returns
  facts, never authorization: `{:ok, :verified}` when any key supplied for
  the entry's `key_id` verifies the envelope; `{:error,
  :signature_key_unsupported}` when no supplied key carries that id;
  `{:error, :signature_algorithm_unsupported}` when the header names a
  non-EdDSA algorithm or no matched key is usable Ed25519;
  `{:error, :signature_malformed}` for envelope violations;
  `{:error, :signature_not_verified}` when the bytes do not verify.
  """
  @spec verify(Json.value(), [PublicKey.t()]) :: {:ok, :verified} | {:error, reason()}
  def verify(entry, keys) when is_list(keys) do
    with {:ok, parts} <- parts(entry, :signature) do
      check(parts, parts.header_b64 <> "." <> parts.payload, keys)
    end
  end

  @doc """
  Verify an attestation entry (identical envelope; `kind` and
  `statement_digest` replace `purpose` and `content_digest`). The kind
  registry is empty, so a well-formed attestation denies
  `{:error, :attestation_malformed}` today — the fail-closed posture, not
  a defect. Shape errors use `:attestation_malformed`; algorithm denials
  (`alg`/`algorithm` not EdDSA/Ed25519) share the signature vocabulary by
  design — the envelope's algorithm posture is one concern. When the first
  kind registers, the single-clause match below fails loudly on the
  now-possible `{:ok, _}` and the identical `check/3` wiring used by
  `verify/2` is added deliberately; the envelope machinery (`parts/2`) is
  already shared.
  """
  @spec verify_attestation(Json.value(), [PublicKey.t()]) :: {:ok, :verified} | {:error, reason()}
  def verify_attestation(entry, _keys) do
    case parts(entry, :attestation) do
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  The RFC 7515 detached compact serialization
  (`BASE64URL(JCS(protected)) .. BASE64URL(signature)`, empty payload
  segment per Appendix F) for standard-tooling cross-checks.
  """
  @spec to_compact(Json.value()) :: {:ok, binary()} | {:error, reason()}
  def to_compact(entry) do
    with {:ok, parts} <- parts(entry, :signature) do
      {:ok, parts.header_b64 <> "." <> "." <> Base64Url.encode(parts.signature)}
    end
  end

  # ---- shared engine ------------------------------------------------------------

  defp parts(entry, shape) do
    with {:ok, members} <- entry_members(entry, shape),
         {:ok, header} <- parse_header(members["protected"], shape),
         {:ok, attrs} <- parse_attrs(members["signed_attributes"], shape),
         :ok <- kid_matches(attrs, header, shape),
         {:ok, signature} <- parse_signature(members["signature"], shape),
         {:ok, header_b64, payload} <-
           canonical_input(members["protected"], members["signed_attributes"]) do
      attributes = if shape == :signature, do: attributes_struct(attrs), else: nil

      {:ok,
       %{
         header_b64: header_b64,
         payload: payload,
         signature: signature,
         key_id: elem(attrs["key_id"], 1),
         attributes: attributes
       }}
    else
      {:error, reason}
      when reason in [
             :signature_malformed,
             :attestation_malformed,
             :signature_algorithm_unsupported
           ] ->
        {:error, reason}

      # canonical_input's bare :error — a member value canonical
      # serialization rejects (e.g. invalid UTF-8 reaching a hand-built
      # entry; decoded entries are UTF-8-clean).
      :error ->
        {:error, malformed(shape)}
    end
  end

  defp entry_members({:object, members}, shape) do
    names = Enum.map(members, &elem(&1, 0))

    if length(names) == 3 and Enum.sort(names) == Enum.sort(@entry_members),
      do: {:ok, Map.new(members)},
      else: {:error, malformed(shape)}
  end

  defp entry_members(_not_an_object, shape), do: {:error, malformed(shape)}

  defp malformed(:signature), do: :signature_malformed
  defp malformed(:attestation), do: :attestation_malformed

  defp parse_header({:object, members}, shape) do
    names = Enum.map(members, &elem(&1, 0))

    if length(names) == 4 and Enum.sort(names) == Enum.sort(@header_members) do
      header = Map.new(members)

      with :ok <- alg_eddsa(header["alg"]),
           :ok <- exact_false(header["b64"]),
           :ok <- exact_crit(header["crit"]),
           {:ok, _kid} <- nonempty_string(header["kid"]) do
        {:ok, header}
      else
        {:error, :signature_algorithm_unsupported} = unsupported -> unsupported
        _other -> {:error, malformed(shape)}
      end
    else
      {:error, malformed(shape)}
    end
  end

  defp parse_header(_not_an_object, shape), do: {:error, malformed(shape)}

  defp alg_eddsa({:string, "EdDSA"}), do: :ok
  defp alg_eddsa({:string, _other}), do: {:error, :signature_algorithm_unsupported}
  defp alg_eddsa(_), do: :error

  defp exact_false({:boolean, false}), do: :ok
  defp exact_false(_), do: :error

  defp exact_crit({:array, [{:string, "b64"}]}), do: :ok
  defp exact_crit(_), do: :error

  defp nonempty_string({:string, s}) when s != "", do: {:ok, s}
  defp nonempty_string(_), do: :error

  defp parse_attrs({:object, members}, shape) do
    expected =
      if shape == :signature, do: @signature_attr_members, else: @attestation_attr_members

    names = Enum.map(members, &elem(&1, 0))

    if length(names) == 5 and Enum.sort(names) == Enum.sort(expected) do
      attrs = Map.new(members)

      with :ok <- algorithm_ed25519(attrs["algorithm"]),
           :ok <- tagged_digest(attrs["content_digest"] || attrs["statement_digest"]),
           :ok <- z_form_timestamp(attrs["created_at"]),
           :ok <- dot_free_key_id(attrs["key_id"]),
           :ok <- closed_member(attrs["purpose"] || attrs["kind"], shape) do
        {:ok, attrs}
      else
        {:error, :signature_algorithm_unsupported} = unsupported -> unsupported
        _other -> {:error, malformed(shape)}
      end
    else
      {:error, malformed(shape)}
    end
  end

  defp parse_attrs(_not_an_object, shape), do: {:error, malformed(shape)}

  defp algorithm_ed25519({:string, "Ed25519"}), do: :ok
  defp algorithm_ed25519({:string, _other}), do: {:error, :signature_algorithm_unsupported}
  defp algorithm_ed25519(_), do: :error

  defp tagged_digest({:string, tagged}) do
    case Digest.from_tagged(tagged) do
      {:ok, _digest} -> :ok
      {:error, _reason} -> :error
    end
  end

  defp tagged_digest(_), do: :error

  defp z_form_timestamp({:string, stamped}) do
    if Regex.match?(@created_at, stamped) and match?({:ok, _, _}, DateTime.from_iso8601(stamped)),
      do: :ok,
      else: :error
  end

  defp z_form_timestamp(_), do: :error

  defp dot_free_key_id({:string, key_id}) when key_id != "" do
    if String.contains?(key_id, "."), do: :error, else: :ok
  end

  defp dot_free_key_id(_), do: :error

  defp closed_member({:string, value}, :signature) do
    if Map.has_key?(@purposes, value), do: :ok, else: :error
  end

  defp closed_member({:string, value}, :attestation) do
    if Map.has_key?(@kinds, value), do: :ok, else: :error
  end

  defp closed_member(_, _), do: :error

  defp kid_matches(attrs, header, shape) do
    {:string, kid} = header["kid"]
    {:string, key_id} = attrs["key_id"]

    if kid == key_id, do: :ok, else: {:error, malformed(shape)}
  end

  defp parse_signature({:string, body}, shape) do
    case Base64Url.decode(body) do
      {:ok, <<_::512>> = signature} -> {:ok, signature}
      {:ok, _wrong_length} -> {:error, malformed(shape)}
      {:error, _reason} -> {:error, malformed(shape)}
    end
  end

  defp parse_signature(_, shape), do: {:error, malformed(shape)}

  defp canonical_input(header, attrs) do
    # RFC 7797 5.2: an unencoded payload containing "." can never appear in
    # a compact serialization. The dot-free key_id validation makes that
    # invariant hold by construction; every other member is dot-free from
    # its closed shape. (A belt re-check here would be an unreachable
    # branch — the coverage contract keeps guards live, not decorative.)
    with {:ok, header_json} <- wrap_encode(header),
         {:ok, attrs_json} <- wrap_encode(attrs) do
      {:ok, Base64Url.encode(header_json), attrs_json}
    end
  end

  defp wrap_encode(value) do
    case Canonicalization.encode(value) do
      {:ok, json} -> {:ok, json}
      {:error, _reason} -> :error
    end
  end

  defp attributes_struct(attrs) do
    {:string, purpose} = attrs["purpose"]

    %Attributes{
      algorithm: "Ed25519",
      content_digest: elem(Digest.from_tagged(elem(attrs["content_digest"], 1)), 1),
      created_at: elem(attrs["created_at"], 1),
      key_id: elem(attrs["key_id"], 1),
      purpose: @purposes[purpose]
    }
  end

  defp check(parts, input, keys) do
    candidates = for %PublicKey{key_id: id} = key <- keys, id == parts.key_id, do: key

    ed25519_keys =
      Enum.filter(candidates, fn
        %PublicKey{algorithm: :ed25519, key: <<encoding::binary-size(32)>>} ->
          usable_ed25519_key?(encoding)

        _not_usable_ed25519 ->
          false
      end)

    cond do
      candidates == [] ->
        {:error, :signature_key_unsupported}

      ed25519_keys == [] ->
        {:error, :signature_algorithm_unsupported}

      Enum.any?(ed25519_keys, fn key ->
        :crypto.verify(:eddsa, :none, input, parts.signature, [key.key, :ed25519])
      end) ->
        {:ok, :verified}

      true ->
        {:error, :signature_not_verified}
    end
  end

  # ---- small-order public-key rejection -------------------------------------------
  # Confirmed live on this OTP: :crypto.verify
  # ACCEPTS a small-order public key with an all-zero signature — a universal
  # forgery whenever such a key enters a trust set (a zeroed key is exactly
  # the "key failed to load" misconfiguration). A supplied key is usable
  # Ed25519 only if it decodes to a canonical on-curve point of the main
  # (large) subgroup. The test is the algebraic characterization of the
  # 8-torsion (as libsodium's ge25519_has_small_order computes it): x = 0,
  # or y = 0, or y*sqrt(-1) = +/- x. Field constants per RFC 8032 5.1;
  # the conformance suite recomputes d and sqrt(-1) from the curve
  # definition and asserts these literals.

  @ed_p 57_896_044_618_658_097_711_785_492_504_343_953_926_634_992_332_820_282_019_728_792_003_956_564_819_949
  @ed_d 37_095_705_934_669_439_343_138_083_508_754_565_189_542_113_879_843_219_016_388_785_533_085_940_283_555
  @ed_sqrt_m1 19_681_161_376_707_505_956_807_079_304_988_542_015_446_066_515_923_890_162_744_021_073_123_829_784_752

  defp usable_ed25519_key?(<<encoding::binary-size(32)>>) do
    <<y_raw::little-unsigned-256>> = encoding
    # RFC 8032 5.1.2: bit 255 carries the sign of x; the low 255 bits are y.
    sign_x = Bitwise.band(y_raw, Bitwise.bsl(1, 255)) != 0
    y = Bitwise.band(y_raw, Bitwise.bsl(1, 255) - 1)

    if y >= @ed_p do
      # Non-canonical y — not a valid encoding.
      false
    else
      yy = Integer.mod(y * y, @ed_p)
      denominator = Integer.mod(@ed_d * yy + 1, @ed_p)

      x2 = Integer.mod((y * y - 1) * mod_pow(denominator, @ed_p - 2), @ed_p)

      # p = 2^255 - 19 is 5 mod 8: the square root is z = a^((p+3)/8),
      # corrected by sqrt(-1) when z^2 = -a. (The a^((p+1)/4) shortcut is
      # INVALID for this p — a mistake this comment records so it never
      # returns.)
      z = mod_pow(x2, div(@ed_p + 3, 8))

      root =
        cond do
          Integer.mod(z * z - x2, @ed_p) == 0 -> z
          Integer.mod(z * z + x2, @ed_p) == 0 -> Integer.mod(z * @ed_sqrt_m1, @ed_p)
          true -> nil
        end

      case root do
        nil ->
          # (y^2 - 1)/(d*y^2 + 1) is not a square — not on the curve.
          false

        0 ->
          # root == 0 happens iff y ∈ {1, p-1} — the identity (order 1) and
          # the order-2 point. The sign-set encodings are non-canonical
          # (RFC 8032 5.1.2) and the sign-clear ones are canonical but
          # small-order; BOTH reject. (Found by the second-language
          # verifier's torsion battery: the previous `not sign_x` admitted the
          # sign-clear half — the exact forgery class this guard exists for.)
          false

        root ->
          not small_order?(canonical_x(root, sign_x), y)
      end
    end
  end

  defp canonical_x(root, sign_x) do
    root_odd = Bitwise.band(root, 1) == 1

    if sign_x == root_odd,
      do: root,
      else: @ed_p - root
  end

  defp small_order?(x, y) do
    y_sqrt_m1 = Integer.mod(y * @ed_sqrt_m1, @ed_p)

    x == 0 or y == 0 or y_sqrt_m1 == x or y_sqrt_m1 == @ed_p - x
  end

  defp mod_pow(base, exponent) do
    mod_pow_loop(base, exponent, 1)
  end

  defp mod_pow_loop(_base, 0, acc), do: acc

  defp mod_pow_loop(base, exponent, acc) do
    acc = if Bitwise.band(exponent, 1) == 1, do: Integer.mod(acc * base, @ed_p), else: acc
    mod_pow_loop(Integer.mod(base * base, @ed_p), Bitwise.bsr(exponent, 1), acc)
  end
end

File: verifier/signature.ts
// Detached JWS verification — the mirror of AgentBlueprintProtocol.Signature:
// RFC 7515 compact + RFC 7797 b64=false unencoded detached payload, Ed25519
// via node:crypto (one-shot verify arg order on node ≥ 24 is
// (algorithm, data, key, signature) — probed on v25.9.0), and the
// small-order / off-curve public-key rejection the twin added after the
// found live (:crypto AND OpenSSL both accept small-order keys
// with all-zero signatures — a universal forgery).
//
// Attestations (verify_attestation) are not mirrored: no corpus surface
// dispatches them and the kind registry is empty by design.

import { createPublicKey, verify as cryptoVerify, type KeyObject } from "node:crypto";
import * as b64 from "./b64url.ts";
import { encode } from "./canonical.ts";
import * as digest from "./digest.ts";
import type { Value } from "./value.ts";
import { member, memberString } from "./corpus.ts";

const PURPOSES = new Set(["blueprint", "deployment", "federation-envelope"]);
const CREATED_AT = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;

export interface PublicKey {
  keyId: string;
  key: Buffer; // raw 32-byte Ed25519 encoding
}

type VerifyResult = { ok: true; v: "verified" } | { ok: false; e: string };

function malformed(): VerifyResult {
  return { ok: false, e: "signature_malformed" };
}

export function verify(entry: Value, keys: PublicKey[]): VerifyResult {
  const parts = parseParts(entry);
  if (!parts.ok) return parts;
  const input = parts.v.headerB64 + "." + parts.v.payload;
  return check(input, parts.v.signature, parts.v.keyId, keys);
}

export function signingInput(entry: Value): { ok: true; v: string } | VerifyResult {
  const parts = parseParts(entry);
  if (!parts.ok) return parts;
  return { ok: true, v: parts.v.headerB64 + "." + parts.v.payload };
}

// RFC 7515 detached compact: BASE64URL(JCS(protected)) .. BASE64URL(sig),
// empty payload segment (Appendix F).
export function toCompact(entry: Value): { ok: true; v: string } | VerifyResult {
  const parts = parseParts(entry);
  if (!parts.ok) return parts;
  return { ok: true, v: parts.v.headerB64 + "." + "." + b64.encode(parts.v.signature) };
}

interface Parts {
  headerB64: string;
  payload: string;
  signature: Buffer;
  keyId: string;
}

function parseParts(entry: Value): { ok: true; v: Parts } | VerifyResult {
  if (entry.t !== "obj") return malformed();
  const names = entry.v.map(([k]) => k).sort();
  if (
    names.length !== 3 ||
    names[0] !== "protected" ||
    names[1] !== "signature" ||
    names[2] !== "signed_attributes"
  ) {
    return malformed();
  }

  const header = member(entry, "protected")!;
  const attrs = member(entry, "signed_attributes")!;

  // Header: exactly {alg, b64, crit, kid}; alg must be EdDSA (the twin's one
  // algorithm-unsupported denial in the header), b64 exactly false, crit
  // exactly ["b64"], kid a non-empty string.
  if (header.t !== "obj") return malformed();
  const headerNames = header.v.map(([k]) => k).sort();
  if (headerNames.length !== 4 || headerNames.join() !== "alg,b64,crit,kid") {
    return malformed();
  }
  const alg = member(header, "alg");
  if (alg !== null && alg.t === "str" && alg.v !== "EdDSA") {
    return { ok: false, e: "signature_algorithm_unsupported" };
  }
  if (alg === null || alg.t !== "str") return malformed();
  const b64Flag = member(header, "b64");
  if (b64Flag === null || b64Flag.t !== "bool" || b64Flag.v !== false) return malformed();
  const crit = member(header, "crit");
  if (
    crit === null ||
    crit.t !== "arr" ||
    crit.v.length !== 1 ||
    crit.v[0]!.t !== "str" ||
    crit.v[0]!.v !== "b64"
  ) {
    return malformed();
  }
  const kid = member(header, "kid");
  if (kid === null || kid.t !== "str" || kid.v === "") return malformed();

  // Attributes: exactly the five closed members.
  if (attrs.t !== "obj") return malformed();
  const attrNames = attrs.v.map(([k]) => k).sort();
  if (
    attrNames.length !== 5 ||
    attrNames.join() !== "algorithm,content_digest,created_at,key_id,purpose"
  ) {
    return malformed();
  }
  const algorithm = member(attrs, "algorithm");
  if (algorithm !== null && algorithm.t === "str" && algorithm.v !== "Ed25519") {
    return { ok: false, e: "signature_algorithm_unsupported" };
  }
  if (algorithm === null || algorithm.t !== "str") return malformed();

  const contentDigest = memberString(attrs, "content_digest");
  if (contentDigest === null || !digest.fromTagged(contentDigest).ok) return malformed();

  const createdAt = memberString(attrs, "created_at");
  if (createdAt === null || !CREATED_AT.test(createdAt) || !validZFormDate(createdAt)) {
    return malformed();
  }

  const keyId = memberString(attrs, "key_id");
  if (keyId === null || keyId === "" || keyId.includes(".")) return malformed();

  const purpose = memberString(attrs, "purpose");
  if (purpose === null || !PURPOSES.has(purpose)) return malformed();

  if (kid.v !== keyId) return malformed();

  const signatureBody = member(entry, "signature");
  if (signatureBody === null || signatureBody.t !== "str") return malformed();
  const decodedSignature = b64.decodeStrict(signatureBody.v);
  if (!decodedSignature.ok || decodedSignature.v.length !== 64) return malformed();

  const headerJson = encode(header);
  const attrsJson = encode(attrs);
  if (!headerJson.ok || !attrsJson.ok) return malformed();

  return {
    ok: true,
    v: {
      headerB64: b64.encode(Buffer.from(headerJson.v, "utf8")),
      payload: attrsJson.v,
      signature: decodedSignature.v,
      keyId,
    },
  };
}

// DateTime.from_iso8601's field validation for the Z whole-second form:
// regex passes still fail on impossible dates (2026-02-31).
function validZFormDate(stamped: string): boolean {
  const year = Number(stamped.slice(0, 4));
  const month = Number(stamped.slice(5, 7));
  const day = Number(stamped.slice(8, 10));
  const hour = Number(stamped.slice(11, 13));
  const minute = Number(stamped.slice(14, 16));
  const second = Number(stamped.slice(17, 19));
  if (month < 1 || month > 12 || day < 1 || hour > 23 || minute > 59 || second > 59) {
    return false;
  }
  const leap = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const lengths = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  return day <= lengths[month - 1]!;
}

function check(
  input: string,
  signature: Buffer,
  keyId: string,
  keys: PublicKey[],
): VerifyResult {
  const candidates = keys.filter((key) => key.keyId === keyId);
  if (candidates.length === 0) return { ok: false, e: "signature_key_unsupported" };

  const usable = candidates.filter(
    (key) => key.key.length === 32 && usableEd25519Key(key.key),
  );
  if (usable.length === 0) return { ok: false, e: "signature_algorithm_unsupported" };

  const inputBytes = Buffer.from(input, "utf8");
  for (const key of usable) {
    if (cryptoVerify(null, inputBytes, keyObject(key.key), signature)) {
      return { ok: true, v: "verified" };
    }
  }
  return { ok: false, e: "signature_not_verified" };
}

const keyObjectCache = new Map<string, KeyObject>();
function keyObject(raw: Buffer): KeyObject {
  const x = raw.toString("base64url");
  const cached = keyObjectCache.get(x);
  if (cached !== undefined) return cached;
  const key = createPublicKey({ key: { kty: "OKP", crv: "Ed25519", x }, format: "jwk" });
  keyObjectCache.set(x, key);
  return key;
}

// ---- small-order / off-curve rejection (RFC 8032 §5.1 algebra in BigInt) --------

const ED_P = 2n ** 255n - 19n;
const ED_D =
  37095705934669439343338083508754565189542113879843219016388785533085940283555n;
const ED_SQRT_M1 =
  19681161376707505956807079304988542015446066515923890162744021073123829784752n;

function modPow(base: bigint, exponent: bigint): bigint {
  let result = 1n;
  let b = base % ED_P;
  let e = exponent;
  while (e > 0n) {
    if (e & 1n) result = (result * b) % ED_P;
    b = (b * b) % ED_P;
    e >>= 1n;
  }
  return result;
}

export function usableEd25519Key(encoding: Buffer): boolean {
  if (encoding.length !== 32) return false;
  const yRaw = encoding.readBigUInt64LE(0) | (encoding.readBigUInt64LE(8) << 64n) |
    (encoding.readBigUInt64LE(16) << 128n) | (encoding.readBigUInt64LE(24) << 192n);
  // RFC 8032 5.1.2: bit 255 carries the sign of x; the low 255 bits are y.
  const signX = ((yRaw >> 255n) & 1n) === 1n;
  const y = yRaw & ((1n << 255n) - 1n);

  if (y >= ED_P) return false; // non-canonical y

  const yy = (y * y) % ED_P;
  const denominator = (ED_D * yy + 1n) % ED_P;
  const x2 = ((yy - 1n) * modPow(denominator, ED_P - 2n)) % ED_P;

  // p = 2^255 - 19 is 5 mod 8: sqrt is a^((p+3)/8), corrected by sqrt(-1)
  // when z^2 = -a. (a^((p+1)/4) is INVALID for this p.)
  const z = modPow(x2, (ED_P + 3n) / 8n);

  let root: bigint | null;
  if ((z * z - x2) % ED_P === 0n) root = z;
  else if ((z * z + x2) % ED_P === 0n) root = (z * ED_SQRT_M1) % ED_P;
  else root = null;

  if (root === null) return false; // not on the curve
  // root == 0 happens iff y ∈ {1, p−1} — the identity (order 1) and the
  // order-2 point: non-canonical when sign-set, small-order when sign-clear;
  // both reject (mirrors the Elixir fix found by this mirror's torsion battery).
  if (root === 0n) return false;

  const rootOdd = root % 2n === 1n;
  const x = signX === rootOdd ? root : ED_P - root;
  return !smallOrder(x, y);
}

function smallOrder(x: bigint, y: bigint): boolean {
  const ySqrtM1 = (y * ED_SQRT_M1) % ED_P;
  return x === 0n || y === 0n || ySqrtM1 === x || ySqrtM1 === ED_P - x;
}
